File: Tecniques de Programació Avançades/Practica 2 - Aleix Carrillo i Oriol Poblet/src/main/java/UniformGroup.scala
class UniformGroup[U, V](var groupSize: Int) extends PolicyManager[U, V] {
  /**
   * Método que devuelve un invoker que cumple con los requisitos de memoria
   *
   * @param invokers       Lista de invokers
   * @param requiredMemory Memoria necesaria para la ejecución
   * @return Invoker que cumple con los requisitos de memoria
   */
  override def valorsInvoker(invokers: List[Invoker[U, V]], requiredMemory: Int): Invoker[U, V] = {
    val totalInvokers = invokers.size
    var remainingMemory = requiredMemory

    for (i <- 0 until totalInvokers if remainingMemory > 0) {
      val invoker = invokers(i)
      val assignedMemory = Math.min(groupSize, remainingMemory)

      if (assignedMemory > 0) {
        remainingMemory -= assignedMemory
        invoker.liberarMemoria(assignedMemory)
        return invoker
      }
    }

    throw new NoSuchElementException("No hay Invokers disponibles")
  }
}

File: Tecniques de Programació Avançades/Practica 1/src/UniformGroup.java
import java.util.List;
import java.util.NoSuchElementException;

public class UniformGroup implements PolicyManager {

    private int groupSize;

    /**
     * Politica Uniform Group: envia de forma uniforme a los invokers los datos a ejecutar
     * @param groupSize Tamaño del grupo
     */

    public UniformGroup(int groupSize) {
        this.groupSize = groupSize;
    }

    /**
     * Método que devuelve un invoker que cumple con los requisitos de memoria
     * @param invokers Lista de invokers
     * @param requiredMemory Memoria necesitada para la ejecución
     * @return Invoker que cumple con los requisitos de memoria
     */

    @Override
    public Invoker valorsInvoker(List<Invoker> invokers, int requiredMemory) {
        int totalInvokers = invokers.size();
        int functionsPerInvoker = requiredMemory / groupSize;

        for (int i = 0; i < totalInvokers && requiredMemory > 0; i++) {
            Invoker invoker = invokers.get(i);
            int assignedMemory = Math.min(functionsPerInvoker, invoker.getAvailableMemory());

            if (assignedMemory > 0) {
                requiredMemory -= assignedMemory;
                invoker.liberarMemoria(assignedMemory);
                return invoker;
            }
        }

        throw new NoSuchElementException("No hay Invokers disponibles");
    }

    /**
     * Devuelve el tamaño del grupo
     * @return groupSize
     */

    public int getGroupSize() {
        return groupSize;
    }
}
